### PR TITLE
Benchmarks: update to latest libraries and add tests

### PR DESCRIPTION
**Pull request by NickCraver**
This builds on the work @mathewc started with benchmarks. There are a lot of things that would be nice to have in benchmark form as we evaluate improving various parts of pipeline performance like async/await elimination, `Task` vs. `ValueTask`, allocations, general algorithmic improvements, etc.

Overall:
- Upgrades to latest BenchmarkDotNet
- Moves the benchmarks project out of tests/ to benchmarks/ (this allows us to do fun things with `Directory.Build.props` and such down the road for _only_ tests and simplifies some things - the move itself is minor just doing it up front.
- Adds some benchmarks as examples - a few areas I'm poking at but haven't gotten into PRs yet (these can be dropped from this PR if wanted).

To run benchmarks:
```ps1
dotnet run -c Release -f net6.0 --project .\benchmarks\WebJobs.Script.Benchmarks\
```

This will present a prompt with all benchmarks:
```text
Available Benchmarks:
  #0 AuthUtilityBenchmarks
  #1 CSharpCompilationBenchmarks
  #2 ScriptLoggingBuilderExtensionsBenchmarks

You should select the target benchmark(s). Please, print a number of a benchmark (e.g. `0`) or a contained benchmark caption (e.g. `AuthUtilityBenchmarks`).
If you want to select few, please separate them with space ` ` (e.g. `1 2 3`).
You can also provide the class name in console arguments by using --filter. (e.g. `--filter *AuthUtilityBenchmarks*`).
```
Note the instructions to do similar from the command line, e.g.:
```ps1
dotnet run -c Release -f net6.0 --project .\benchmarks\WebJobs.Script.Benchmarks\ --filter *ScriptLogging*
```

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

I think these are all infrastructure/help changes and don't need the above, but happy to make any additions or changes we should.

### CODE REVIEW

**Comment by liliankasem** [comment]
Do you mind adding a README.md in the `benchmarks/` folder with similar instructions to what you have in the PR desc?

File: benchmarks/WebJobs.Script.Benchmarks/CSharpCompilationBenchmarks.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        // Set of samples to benchmark
+        // TODOO: BlobTrigger, needs assembly refs working
+        [Params("DocumentDB", "HttpTrigger", "HttpTrigger-CustomRoute", "HttpTrigger-Weather", "NotificationHub")]
+        public string BenchmarkTrigger;
+
+        // Script source
+        private string ScriptPath;
+        private string GetCSharpSamplePath([CallerFilePath] string thisFilePath = null) =>
+            Path.Combine(thisFilePath, "../../../sample/CSharp");
```
Comment (liliankasem) [inline benchmarks/WebJobs.Script.Benchmarks/CSharpCompilationBenchmarks.cs]: Can we replace the `/` with `"..", "..", ".."` so it's also mac friendly :)

Comment (NickCraver) [reply benchmarks/WebJobs.Script.Benchmarks/CSharpCompilationBenchmarks.cs]: oo good catch - yes!

File: benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            var job = Job.Default
+                .WithWarmupCount(1) // 1 warmup is enough for our purpose
+                .WithIterationTime(TimeInterval.FromMilliseconds(250)) // the default is 0.5s per iteration, which is slighlty too much for us
+                .WithMinIterationCount(15)
+                .WithMaxIterationCount(20); // we don't want to run more that 20 iterations
+
+            return DefaultConfig.Instance
+                .AddLogger(ConsoleLogger.Default) // log output to console
+                .AddExporter(MarkdownExporter.GitHub) // export to GitHub markdown
```
Comment (liliankasem) [inline benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs:33]: Pretty cool that it has a GitHub exporter, I wonder if we would want to utilise this in a GitHub Action so we can do something like `/benchmark` or is the plan to eventually run these in CI at all?

File: benchmarks/WebJobs.Script.Benchmarks/AuthUtilityBenchmarks.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                    if (level == requiredLevel && (keyName == null || string.Equals(principal.FindFirstValue(SecurityConstants.AuthLevelKeyNameClaimType), keyName, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
```
Comment (liliankasem) [inline benchmarks/WebJobs.Script.Benchmarks/AuthUtilityBenchmarks.cs:90]: nit: extra line?

File: benchmarks/WebJobs.Script.Benchmarks/AuthUtilityBenchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            new Claim(SecurityConstants.AuthLevelKeyNameClaimType, "test1"),
+            new Claim(SecurityConstants.AuthLevelKeyNameClaimType, "test2"),
+            new Claim(SecurityConstants.AuthLevelKeyNameClaimType, "test3"),
+            new Claim(SecurityConstants.AuthLevelClaimType, nameof(AuthorizationLevel.Function)),
+            new Claim(SecurityConstants.AuthLevelClaimType, nameof(AuthorizationLevel.Anonymous)),
+            new Claim(SecurityConstants.AuthLevelClaimType, nameof(AuthorizationLevel.User)),
+            new Claim(SecurityConstants.AuthLevelClaimType, nameof(AuthorizationLevel.Admin)),
+            new Claim(SecurityConstants.AuthLevelClaimType, nameof(AuthorizationLevel.System)),
+
```
Comment (liliankasem) [inline benchmarks/WebJobs.Script.Benchmarks/AuthUtilityBenchmarks.cs]: extra line also here, wonder if the ci will pick these up 🤔

**Review by liliankasem — DISMISSED** [review]
lgtm, just some minor comments!

File: benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs (hunk 2 above)
Comment (NickCraver) [reply benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs:33]: I think plugging these in would be a great next step and absolutely something I'd like to do - but I'm not sure the best (most usable) way. Current intent is to reach out to some people out for the year (e.g. Adam Sitnik) to see what other teams are doing here and advice on what we can best utilize. If you or anyone else has ideas here, maybe we should start an issue on that integration? Was pacing myself on not creating _too_ much noise while people are out :)

**Comment by NickCraver** [comment]
@liliankasem Added a README (and included in the solution) - could use a full review there, happy to edit/add!

File: benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs (hunk 2 above)
Comment (liliankasem) [reply benchmarks/WebJobs.Script.Benchmarks/RecommendedConfig.cs:33]: Created an issue! #7998

**Comment by liliankasem** [comment]
> @liliankasem Added a README (and included in the solution) - could use a full review there, happy to edit/add!

Looks good, thanks! It might be nice to also add instructions on how to create a new benchmark test, if not here then maybe in our [internal docs](https://dev.azure.com/msazure/One/_git/AAPT-Antares-Docs?path=/TeamDocs/FunctionTeamDocs/EngineeringHowTos/toc.yml&version=GBmaster&line=1)

oh and nit for adding a new line before the code blocks (I have all the [markdown tools ](https://marketplace.visualstudio.com/items?itemName=docsmsft.docs-authoring-pack)and so the linting is really in my face 😂)

**Comment by NickCraver** [comment]
@liliankasem I think adding benchmark addition instructions to the README is great - recommend we merge this into your branch and point the PR to be on it instead of dev, then use today to make sure we have good instructions from the first addition :)

And cheers for extension - installing!

**Review by liliankasem — APPROVED** [review]

File: benchmarks/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Benchmarks
+
+Welcome to the benchmarks! This folder is for code benchmarking (e.g. components rather than end-to-do testing).
+The intent is to benchmark areas we think are interesting and measure improvements as well as ensuring we don't unintentionally regress over time.
+
+There's a lot of things that would be nice to have in benchmark form as we evaluate improving various parts of pipeline performance like async/await elimination,
+`Task` vs. `ValueTask`, allocations, general algorithmic improvements, etc. This is where those assessments live.
+
+To run benchmarks (from solution root - otherwise shorten the project path!):
```
Comment (kshyju) [inline benchmarks/README.md:9]: Benchmarks are still going to be comparing 2 method implementations on same build. correct? Not comparing it to the numbers for same method from previous build (I see we don't have any infra to persist that info). Is that something we plan on setting up in future?

**Review by kshyju — APPROVED** [review]

File: benchmarks/README.md (hunk 5 above)
Comment (NickCraver) [reply benchmarks/README.md:9]: We have the flexibility either way, long term: yes I'd like to auto-run and detect regressions (or improvements!), but for the moment what a PR author could do is: run the benchmark against dev and then again with their changes and post the results as an intro to getting this going at all. Stage 1 is making at least that much easy, then we can iterate baking them into the process more. ...not sure I phrased that well though, does it make sense?

Comment (kshyju) [reply benchmarks/README.md:9]: Yes. makes sense. Thank you!

**Review by mathewc — APPROVED** [review]

**Pull request merged**